flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStoreTest.java

```diff
@@ -0,0 +1,382 @@
+					store.add(key, state);
+
+					try {
+						store.add(key, state);
+						fail("Exception should be thrown.");
+					} catch (StateHandleStore.AlreadyExistException ex) {
+						final String msg = String.format(
+							"%s already exists in ConfigMap %s", key, LEADER_CONFIGMAP_NAME);
+						assertThat(ex.getMessage(), containsString(msg));
+					}
+					assertThat(longStateStorage.getStateHandles().size(), is(2));
+					assertThat(longStateStorage.getStateHandles().get(1).getNumberOfDiscardCalls(), is(1));
+				});
+		}};
+	}
+
+	@Test
+	public void testAddFailedWhenConfigMapNotExistAndDiscardState() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					final KubernetesStateHandleStore<Long> store = new KubernetesStateHandleStore<>(
+						flinkKubeClient, LEADER_CONFIGMAP_NAME, longStateStorage, filter, LOCK_IDENTITY);
+
+					try {
+						store.add(key, state);
+						fail("Exception should be thrown.");
+					} catch (Exception ex) {
+						final String msg = String.format("ConfigMap %s does not exist.", LEADER_CONFIGMAP_NAME);
+						assertThat(ex.getMessage(), containsString(msg));
+					}
+					assertThat(longStateStorage.getStateHandles().size(), is(1));
+					assertThat(longStateStorage.getStateHandles().get(0).getNumberOfDiscardCalls(), is(1));
+				});
+		}};
+	}
+
+	@Test
+	public void testReplace() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					leaderCallbackGrantLeadership();
+
+					final KubernetesStateHandleStore<Long> store = new KubernetesStateHandleStore<>(
+						flinkKubeClient, LEADER_CONFIGMAP_NAME, longStateStorage, filter, LOCK_IDENTITY);
+
+					store.add(key, state);
+
+					final Long newState = 23456L;
+					final String resourceVersion = store.exists(key);
+					store.replace(key, resourceVersion, newState);
+
+					assertThat(store.getAll().size(), is(1));
+					assertThat(store.get(key).retrieveState(), is(newState));
+				});
+		}};
+	}
+
+	@Test
+	public void testReplaceWithKeyNotExist() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					leaderCallbackGrantLeadership();
+
+					final KubernetesStateHandleStore<Long> store = new KubernetesStateHandleStore<>(
+						flinkKubeClient, LEADER_CONFIGMAP_NAME, longStateStorage, filter, LOCK_IDENTITY);
+					final Long newState = 23456L;
+
+					try {
+						assertThat(store.exists(key), is(StateHandleStore.NON_EXIST_RESOURCE_VERSION));
+						store.replace(key, StateHandleStore.NON_EXIST_RESOURCE_VERSION, newState);
+						fail("Exception should be thrown.");
+					} catch (StateHandleStore.NotExistException e) {
+						final String msg = String.format(
+							"Could not find %s in ConfigMap %s", key, LEADER_CONFIGMAP_NAME);
+						assertThat(e.getMessage(), containsString(msg));
+					}
+					assertThat(store.getAll().size(), is(0));
+				});
+		}};
+	}
+
+	@Test
+	public void testReplaceFailedAndDiscardState() throws Exception {
```

Comment:
Actually, when we do not have the leadership, `checkAndUpdate` will return false in `KubernetesStateHandle#replace`. I assume you mean to throw exception when trying to update the ConfigMap.

Reply:
It is `KubernetesStateHandleStore`'s logic that it will give up updating the config map if it does not have leadership.
I meant we should also verify the case that `KubernetesStateHandleStore` decides to update the config map but the update fails.

Reply:
The return value of `checkAndUpdate` is false does not mean that we update the ConfigMap failed. It means we do not update the ConfigMap since the `function` return `Optional.empty()`.

When the kube client failed to update the ConfigMap, we will retry and thrown an exception finally. I have added a case for this.

Reply:
Ok, thanks for the explaination.